Add CRFCut sentence segmentation

cstorm125 opened the pull request:
CRFCut -- Thai sentence segmentation with conditional random field, default trained on TED dataset

- ORCHID - space-correct accuracy 87% vs 95% state-of-the-art (Zhou et al, 2016; https://www.aclweb.org/anthology/C16-1031.pdf)
- TED dataset - space-correct accuracy 82%

See development notebooks at https://github.com/vistec-AI/ted_crawler;
POS features are not used due to unreliable POS tagging available

Commit:
crfcut and tests

pep8speaks commented:
Hello @cstorm125! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:













There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2019-12-20 17:27:11 UTC

bact committed:
Update test_tokenize.py

bact committed:
Update __init__.py

bact committed:
refactor crfcut.py

collapse two for loop to one

bact committed:
Update crfcut.py

wannaphong committed:
Update .travis.yml

wannaphong committed:
Update .travis.yml

wannaphong committed:
Update .travis.yml

wannaphong committed:
Close wordnet test

wannaphong committed:
Update test_corpus.py

bact committed:
Update .travis.yml

bact committed:
Update test_corpus.py

bact committed:
Update test_corpus.py

bact committed:
Add pytest-xdist

bact committed:
coverage run --concurrency=multiprocessing

bact committed:
Update .travis.yml

bact committed:
Merge branch 'dev' into dev

bact commented:
1) `python-crfsuite` should be included in `setup.py`'s `install_requires` (and maybe in `requirements.txt` as well, see next point)
  - The code currently passed the test only because in the test the `python-crfsuite` was installed as dependencies of `attacut` - but in real settings, user may not install the optional `attacut`.

2) How big is the model? If it's not that big I would suggest to include it in the module, so we will less dependent on the network during runtime.
  - This will make it possible to have `crfcut` included as a pythainlp core, as a default engine for sentence segmentation.
  - If this is really the case, `python-crfsuite` should be in `requirements.txt` too.
  - The model file should be included in `setup.py`'s `package_data`
  - Once the model included, `_download()` function should be removed.

cstorm125 commented:
@bact The model is 5MB so I agree we can include it as battery. Is everyone okay with having some model files in the library? @artificiala @wannaphongcom

bact commented:
For reference on model size, see https://github.com/PyThaiNLP/pythainlp/issues/298 

I think 5 MB is ok.

wannaphong commented on pythainlp/tokenize/__init__.py:
```diff
@@ -179,17 +179,17 @@ def dict_word_tokenize(
     :return: list of splited sentences
     :rtype: list[str]
     **Options for engine**
+        * *crfcut* - split by CRF trained on TED dataset
```
`* *crfcut* - split by CRF trained on TED dataset` -> `* *crfcut* (default) - split by CRF trained on TED dataset`

wannaphong commented on pythainlp/tokenize/__init__.py:
```diff
@@ -179,17 +179,17 @@ def dict_word_tokenize(
     :rtype: list[str]
     **Options for engine**
+        * *crfcut* - split by CRF trained on TED dataset
         * *whitespace+newline* (default) - split by whitespace token \
```
`* *whitespace+newline* (default) - split by whitespace token` -> `* *whitespace+newline* - split by whitespace token`

bact committed:
Update __init__.py

bact committed:
Update test_tokenize.py

Commit:
add python-crfsuite to requirements.txt and setup.py

Commit:
change crfcut to battery-included

Commit:
import os in crfcut.py

bact committed:
Update __init__.py

bact committed:
Add starter and ender words, change list to set

- add few words to the STARTERS and ENDERS lists
- change word list to set, faster membership test

bact committed:
uncommeted wordnet tests

bact committed:
Format test_tokenize.py

bact committed:
Sort requirements.txt

bact committed:
Uncomment wordnet

bact committed:
Sort requirements in setup.py

bact committed:
Add sentenceseg-ted.model to package_data

bact reviewed (changes requested):
I have add few words to STARTERS and ENDERS. May require retrain and reupload of the model.
Apart of that, I think we're good to go.

Please also kindly update the table here: https://github.com/PyThaiNLP/pythainlp/issues/298 thx

Great work! Another step towards full pipeline.

bact committed:
Fix requirements

bact committed:
Merge branch 'dev' into dev

Commit:
update sentenceseg-ted.model with new starters and enders

wannaphong commented:
:100:

bact commented:
TODO: Next step is to convert https://github.com/vistec-AI/ted_crawler/blob/master/sentenceseg_ted.ipynb to a commandline script and maybe put it in `bin/` directory. So people can train their own model. Follow up on #344.
